proposals/infra/INF-0004-register-types-and-diagnostics.md

```diff
@@ -129,97 +137,165 @@ final case will be entered. Types that are or contain a resource are known as
+decl is a UDT, so the `udt` flag is set. We then recurse through the members of the UDT and
+collect information on what types of resources are contained by the UDT (uav, cbv, srv, or
+sampler). The corresponding resouce class flags are set. If the decl is not a struct or class
+type, then it is a basic type, and so the `basic` flag is set.<br>
+The last step is to simply check if the decl is in the `$Globals` scope. If so, then the
+`default_globals` flag will be set.<br><br>
+
+Now enough information has been gathered to determine the right diagnostics to emit, if any.
+Firstly, if `other` is set, then the variable type is not valid for the given register type.
+So, we can emit this diagnostic: `error: '%0' is an invalid resource type for register type '%1'`
+If the `resource` flag is set and the `udt` flag is not set, then we have to check the 
+register statement, and the resource class flag, and verify that the register type matches the 
+resource class flag. If not, we emit this diagnostic: 
+`error; %select{SRV|UAV|CBV|Sampler}2 type '%0' requires register type '%select{t|u|b|s}2', but register type '%1' was used.`
+If the `basic` flag is set, then we first check if `default_globals` is set. If so, then 
+we check the register type. If the register type is 'i' or 'b', the deprecated warning
+that is treated as an error as shown in the examples above will be emitted. If 'c' is given,
+no errors will be emitted. After this point, `default_globals` doesn't need to be set.
+If 't', 'u', or 's' are given, then this diagnostic will be emitted:
+`error: '%0' is an invalid resource type for register type '%1'`. If any other register type
+is seen, this error will be emitted:
+`error: register binding type '%1' not supported for variable of type '%0'`<br><br>
+
+Finally, in the case that `udt` is set, we first check `default_globals`. If it is set,
+then we can permit the 'c' register type. After this point, `default_globals` doesn't
+need to be set. For every register type that is used to bind the resources contained in
+the given UDT, we verify that the corresponding resource class flag has been set. If the
+corresponding resource class flag is not set, this error will be emitted:
+`error: variable of type '%0' bound to register type '%1' does not contain a matching '%select{SRV|UAV|CBV|Sampler}2' resource`
+Otherwise, if any other register type is given, then we emit this error:
+`error: register binding type '%0' not supported for variable of type '%1'`<br><br>
+
 Legacy behavior can be allowed with the --Wno-disallow-legacy-binding-rules
 flag. When this flag is active, and legacy behavior is present, a warning will 
-be emitted instead of an error.
+be emitted instead of an error.<br><br>
 
-Below is a flowchart that describes the process that is used to determine
-what kind of diagnostic to emit in each case:
+Below are two flowcharts that describe the process that is used to determine

```

Comment:
I'm sorry, I'm really not a fan of the flowcharts.  I find them hard to read in the previews: they're so zoomed out I can't read the text, and if I zoom in then I have to use the clunky navigation things that github adds to them to look around.  And they're essentially write-only in the raw md view.

Reply:
![image](https://github.com/microsoft/hlsl-specs/assets/8118402/fc98d803-ddc7-42f9-b724-b70c3f9e09b0)

Reply:
I do think that they're useful since it's difficult to describe this entire process with just words. Lots of conditions, and it's not always clear where exactly if's `end` unless you're extremely verbose. But I'll remove them for now and see if the description is acceptable.